contributors/design-proposals/federated-resource-quota.md

```diff
@@ -0,0 +1,53 @@
+# Federated ResourceQuota Requirements & Design
+@weijinxu, @IrfanUrRehman, @JianhuiZ
+reviewed by  @quinton-hoole, @deepak-vijSummary
+
+In a single Kubernetes cluster there is a concept of Resource Quota that manages the resources by namespace so that user usages can be limited to their fair share of the resources. We want to expand this concept to the federation level so that users can treat Federation Control Plane as a “Single Pane of Glass” for seamlessly managing the Resource Quota  information across all the underlying federated clusters.
+
+## User scenario
+
+As a federation user and an admin of clusters, I want to be able to define an overall resource quota for some specific or all resources for a given federated namespace. The overall resource quota limit should be applicable across clusters and preferably enforced at federation control plane level.
+
+For example, let us say that we’ve specified 100 pods in the resource quota spec in the federation namespace, with multiple underlying federated clusters. Let us assume that 50 pods have been deployed already across the federation in a previous deployment run. In a subsequent deployment run, if a user requests additional 50+ more pods to be deployed, the deployment will be rejected as a whole by the federation control plane before deployment process actually starts. If user requests less than 50 pods, the deployment should succeed as per the deployment preferences.
```

Comment:
@derekwaynecarr thanks for having a look at this.This text actually implies (as later explained in the doc) that the fcp tries to put a check on the object count quota at federation using admission control (just like what a k8s cluster APIserver would do). Actually the example  using pods is misleading, because the pod access as of now is unavailable at fcp. If pods as an api object is exposed from fcp, this example will stand valid. 

@weijinxu I suggest you to update the example with some object for which the api is exposed from fcp (eg replicaset or configmap) and ensure that the quota type as count is explicitly specified somewhere in this text. 

On the other hand it is correct that for the  other controllers which work on replica count (viz deployment, rs), the check cannot happen at admission. As per this design, that check is delegated to federated clusters by creating cluster local resource quota objects (with some reduced limit, which is continually updated based on the quota status observed from all clusters). The cluster local runtime quota enforcement and update of status on those quota objects is done only by local check quota controller(s).  Please let us know if that sounds like a bad idea.